ggml.c

```diff
@@ -13377,13 +13327,13 @@ static void ggml_compute_forward_soft_max_f32(
 
     float * wp = (float *) params->wdata + (nc + CACHE_LINE_SIZE_F32) * ith;
 
-    // when max_bias <= 0.0f, src2 is not used and we default it to src0 to avoid branching
-    ggml_fp16_t * pos_f16 = src2 ? (ggml_fp16_t *) src2->data : src0->data;
-    float       * pos_f32 = src2 ? (float       *) src2->data : src0->data;
-
-    const bool use_f16 = (src1 && src1->type == GGML_TYPE_F16) || (src2 && src2->type == GGML_TYPE_F16);
+    const bool use_f16 = (src1 && src1->type == GGML_TYPE_F16);
 
     for (int i1 = ir0; i1 < ir1; i1++) {
+        // ALiBi
+        const uint32_t h = (i1/ne01)%ne02; // head
+        const float slope = (max_bias > 0.0f) ? h < n_head_log2 ? powf(m0, h + 1) : powf(m1, 2*(h - n_head_log2) + 1) : 1.0f;
```

Comment:
I do not see how this directly work? how does it implement this logic?

```[0, 1, 2, 3], [1, 0, 1, 2]``` 

and the negative slope or negativity of this matrix?

Reply:
We compute the integer matrix here:

https://github.com/ggerganov/llama.cpp/blob/f7055d31c59d93e1f6517830a61a69a83bc5d067/llama.cpp#L10966-L10987

We store it in `KQ_mask`. The `slope` is just the head-specific hyper parameter `m` from the link

Reply:
ah  ok, good thanks!